File: frontend/src/types/skill-execution.ts
export type ExecutionOverallStatus =
  | 'NOT_STARTED'
  | 'RUNNING'
  | 'SUCCESS'
  | 'FAILED'
  | 'STOPPED'
  | 'UNKNOWN'

export type ExecutionPhaseStatus = 'RUNNING' | 'COMPLETED' | 'FAILED' | 'STOPPED'

export interface SkillExecution {
  execution_id: string
  project_id: string
  stage_id: string
  skill_id: string
  skill_name: string
  trigger_action: string
  current_phase: string
  phase_status: ExecutionPhaseStatus
  overall_status: ExecutionOverallStatus
  retry_count: number
  previous_execution_id: string | null
  is_release_skill: boolean
  release_confirmed: boolean
  started_at: string | null
  completed_at: string | null
  created_at: string
}

export interface LogEntry {
  timestamp: string
  level: 'INFO' | 'WARN' | 'ERROR' | 'DEBUG'
  content: string
}

export interface LogQueryResult {
  log_entries: LogEntry[]
  total_count: number
  next_anchor: string | null
}

File: backend/app/models/skill_execution.py
"""SkillExecution ORM model."""

from __future__ import annotations

from datetime import datetime

from sqlalchemy import CheckConstraint, ForeignKey, Integer, String
from sqlalchemy.orm import Mapped, mapped_column

from app.infrastructure.database.base import Base


class SkillExecution(Base):
    """Skill 执行记录表."""

    __tablename__ = "skill_executions"

    execution_id: Mapped[str] = mapped_column(String(36), primary_key=True)
    project_id: Mapped[str] = mapped_column(
        ForeignKey("projects.project_id", ondelete="CASCADE"),
        nullable=False,
    )
    stage_id: Mapped[str] = mapped_column(String(36), nullable=False)
    skill_id: Mapped[str] = mapped_column(String(36), nullable=False)
    skill_name: Mapped[str] = mapped_column(String(128), nullable=False)
    trigger_action: Mapped[str] = mapped_column(
        String(16), nullable=False, default="SINGLE_EXECUTE"
    )
    current_phase: Mapped[str] = mapped_column(
        String(16), nullable=False, default="NONE"
    )
    phase_status: Mapped[str] = mapped_column(
        String(16), nullable=False, default="RUNNING"
    )
    overall_status: Mapped[str] = mapped_column(
        String(16), nullable=False, default="NOT_STARTED"
    )
    retry_count: Mapped[int] = mapped_column(
        Integer, nullable=False, default=0
    )
    previous_execution_id: Mapped[str | None] = mapped_column(
        ForeignKey("skill_executions.execution_id", ondelete="SET NULL"),
        nullable=True,
    )
    is_release_skill: Mapped[bool] = mapped_column(nullable=False, default=False)
    release_confirmed: Mapped[bool] = mapped_column(nullable=False, default=False)
    started_at: Mapped[datetime | None] = mapped_column(nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        nullable=False, default=datetime.utcnow
    )

    __table_args__ = (
        CheckConstraint(
            "trigger_action IN ('SINGLE_EXECUTE', 'BATCH_EXECUTE', 'RETRY')",
            name="ck_exec_trigger_action",
        ),
        CheckConstraint(
            "current_phase IN ('PREP', 'EXEC', 'POST', 'NONE')",
            name="ck_exec_current_phase",
        ),
        CheckConstraint(
            "phase_status IN ('RUNNING', 'COMPLETED', 'FAILED', 'STOPPED')",
            name="ck_exec_phase_status",
        ),
        CheckConstraint(
            "overall_status IN ('NOT_STARTED', 'RUNNING', 'SUCCESS', 'FAILED', 'STOPPED', 'UNKNOWN')",
            name="ck_exec_overall_status",
        ),
        CheckConstraint(
            "retry_count BETWEEN 0 AND 3",
            name="ck_exec_retry_count",
        ),
    )
